### PR TITLE
v2 submission api requires type and type version.

### PR DESCRIPTION
**Pull request by kshakir**
v1 submission api returns an HTTP warning when using older form data names.
All unrecognized api versions return bad statuses.
Added additional specs.

### CODE REVIEW

File: engine/src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 1]
```diff
@@ -180,59 +181,73 @@ trait CromwellApiService {
...
+      bodyPart => bodyPart.toStrict(duration).map(strict => bodyPart.name -> strict.entity.data)
     }.runFold(Map.empty[String, ByteString])((map, tuple) => map + tuple)
 
     def toResponse(workflowId: WorkflowId): WorkflowSubmitResponse = {
       WorkflowSubmitResponse(workflowId.toString, WorkflowSubmitted.toString)
     }
 
-    def askSubmit(command: WorkflowStoreActor.WorkflowStoreActorSubmitCommand): Route = {
+    def askSubmit(command: WorkflowStoreActor.WorkflowStoreActorSubmitCommand, warnings: Seq[String]): Route = {
       // NOTE: Do not blindly coppy the akka-http -to- ask-actor pattern below without knowing the pros and cons.
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/webservice/CromwellApiService.scala]: Also don't blindly copy this spelling of copy.

File: engine/src/main/scala/cromwell/webservice/CromwellApiService.scala [hunk 2]
```diff
@@ -285,16 +300,40 @@ object CromwellApiService {
...
+      Need a quoted string.
+      https://stackoverflow.com/questions/7886782
+
+      Using a poor version of ~~#!
+      https://github.com/akka/akka-http/blob/v10.0.9/akka-http-core/src/main/scala/akka/http/impl/util/Rendering.scala#L206
+       */
+      val quotedString = "\"" + warning.replaceAll("\"","\\\\\"").replaceAll("[\\r\\n]+", " ").trim + "\""
+
+      // https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.46
+      RawHeader("Warning", s"299 cromwell/$cromwellVersion $quotedString")
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/webservice/CromwellApiService.scala]: kinda surprised we have to roll all this ourselves

File: README.md [hunk 3]
```diff
@@ -2587,6 +2587,10 @@ All web server requests include an API version in the url. The current version i
 This endpoint accepts a POST request with a `multipart/form-data` encoded body.  The form fields that may be included are:
 
 * `workflowSource` - *Required* Contains the workflow source file to submit for execution.
+* `workflowType` - *Required* The type of the `workflowSource`, for example "WDL".
```
Comment (danbills) [inline README.md]: ToL can't we default to Wdl if not specified?

File: engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala [hunk 4]
```diff
@@ -56,8 +61,11 @@ object PartialWorkflowSources {
         partialSources.copy(workflowInputsAux = partialSources.workflowInputsAux + (index -> data.utf8String))
       } else if (name == "workflowOptions") {
         partialSources.copy(workflowOptions = Option(data.utf8String))
-      } else if (name == "wdlDependencies" || name == "workflowDependencies") {
-        if (name == "wdlDependencies") deprecationWarning(out = "wdlDependencies", in = "workflowDependencies")
+      } else if (name == "wdlDependencies" && version == 1) {
+        val warning = deprecationWarning(out = "wdlDependencies", in = "workflowDependencies")
+        val warnings = warning +: partialSources.warnings
+        partialSources.copy(zippedImports = Option(data.toArray), warnings = warnings)
+      } else if (name == "workflowDependencies") {
```
Comment (danbills) [inline engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala]: a pattern match on `(name, version)` would go a long way

File: engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala [hunk 5]
```diff
@@ -96,30 +106,54 @@ object PartialWorkflowSources {
...
-  private def deprecationWarning(out: String, in: String): Unit = {
+  private def deprecationWarning(out: String, in: String): String = {
     val warning =
       s"""
          |The '$out' parameter name has been deprecated in favor of '$in'.
          |Support for '$out' will be removed from future versions of Cromwell.
          |Please switch to using '$in' in future submissions.
          """.stripMargin
     log.warn(warning)
+    warning
```
Comment (danbills) [inline engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala]: just remove `warning` declaration

File: engine/src/main/scala/cromwell/webservice/CromwellApiService.scala (hunk 1 above)
Comment (kshakir) [reply engine/src/main/scala/cromwell/webservice/CromwellApiService.scala]: Hehe, was hoping no one noticed that. I'm going to leave it as a "watermark".

File: engine/src/main/scala/cromwell/webservice/CromwellApiService.scala (hunk 2 above)
Comment (kshakir) [reply engine/src/main/scala/cromwell/webservice/CromwellApiService.scala]: I couldn't find anything in the existing [APIs](https://github.com/akka/akka-http/blob/v10.0.9/akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala) or online. If anyone knows something I missed, happy to change this now or later.

File: README.md (hunk 3 above)
Comment (kshakir) [reply README.md]: We _could_ default the type and type-version, but I believe tech-talk discussions said that the point of bumping to `v2` was the requirement that these be specified. If one doesn't want to specify the values, `v1` will still work (for now).

File: engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala (hunk 5 above)
Comment (kshakir) [reply engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala]: Can't remove this afaik. `warning` is used twice, once logged server side on the previous line, then returned here.

File: CHANGELOG.md [hunk 6]
```diff
@@ -2,6 +2,13 @@
 
 ## 30 Release Notes
 
+### Breaking Changes
+
+* All endpoints now require valid version numbers.
+
+* `POST /api/workflows/v2` now requires a `workflowType` and a `workflowTypeVersion`. These are optional for the `v1`
+api. See the [README](https://github.com/broadinstitute/cromwell#post-apiworkflowsversion) for more information.
```
Comment (katevoss) [inline CHANGELOG.md]: ToL does this mean there is an official V2 of all the endpoints, or just `/api/workflows`?

File: README.md (hunk 3 above)
Comment (geoffjentry) [reply README.md]: I missed that TT but I recall having the same initial reaction as @danbills when I was caught up. However after chatting w/ @kshakir I remember my thinking changing to the percentage of users actually constructing URLs by hand being quite low, not sure it's worth it.

File: CHANGELOG.md (hunk 6 above)
Comment (kshakir) [reply CHANGELOG.md]: Metadata v2 already existed (it's the _exact_ same as v1 due to... at this point no good reason). POST /api/workflows, aka workflow submissions, will now have a v2 that behave differently.

Comment (katevoss) [reply CHANGELOG.md]: Huh, TIL!

Comment (geoffjentry) [reply CHANGELOG.md]: Perhaps this should be a TT tomorrow after standup but I don't think we should be having multiple versions of every endpoint. That's not how most sites do their APIs, at least not in my experience. Every endpoint should accept a `v2` parameter, IMO.

File: README.md (hunk 3 above)
Comment (geoffjentry) [reply README.md]: Although if we were to do this we could get away with just keeping `v1` i believe, as then these changes would be backwards compatible

**Comment by kshakir** [comment]
After discussion, replaced by #2586

**Pull request closed without merging**